include/swift/Basic/Located.h

```diff
@@ -0,0 +1,57 @@
+//===--- Located.h - Source Location and Associated Value ----------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Provides a currency data type Located<T> that should be used instead
+// of std::pair<T, SourceLoc>.
+//
+//===----------------------------------------------------------------------===//
+
+
+#ifndef SWIFT_BASIC_LOCATED_H
+#define SWIFT_BASIC_LOCATED_H
+#include "swift/Basic/Debug.h"
+#include "swift/Basic/LLVM.h"
+#include "swift/Basic/SourceLoc.h"
+
+namespace swift {
+
+/// A currency type for keeping track of items which were found in the source code.
+/// Several parts of the compiler need to keep track of a `SourceLoc` corresponding
+/// to an item, in case they need to report some diagnostics later. For example,
+/// the ClangImporter needs to keep track of where imports were originally written.
+/// Located makes it easy to do so while making the code more readable, compared to
+/// using `std::pair`.
+template<typename T>
+struct Located {
+
+  /// The main item whose source location is being tracked.
+  T Item;
+
+  /// The original source location from which the item was parsed.
+  SourceLoc Loc;
+
+  Located(): Item(), Loc() {}
+
+  Located(T Item, SourceLoc loc): Item(Item), Loc(loc) {}
+
+  SWIFT_DEBUG_DUMP;
+  void dump(raw_ostream &os) const;
+
+  template<typename U>
```

Comment:
Nit: LLVM style prefers the ref against the identifier, not the type.

Reply:
This seems fine. For example, you can look at examples in this section: https://llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments

```
bool fooBar(bool Baz, StringRef Quux, std::vector<int> &Result);
```

Reply:
Oh wait, I misread your comment. We are in agreement, but maybe your comment got moved to a different line or something, as there is nothing here.

Reply:
Fixed.